### ISSUE TITLE
Moving the database into a separate repository?

### ISSUE THREAD

**Issue by eszlari** [lensfun/lensfun#1172]
Was moving the database into a separate repository considered? This would make life easier for packagers and for the development process it would make a clearer distinction between library changes and new camera data.

**Comment by junkyardsparkle** [comment]
I think it's worth considering. It would also help with some of the confusion arising from people thinking they need to update their lensfun version just to get new profiles. I don't know if this was the motivation behind some distros recently shipping the development alpha versions, but I guess it would also help prevent things like that.

**Comment by eszlari** [comment]
Of course this repo should provide both, v1 and v2, of the database format.

**Comment by paperdigits** [comment]
@junkyardsparkle  what would we need to do to make this happen?

**Comment by junkyardsparkle** [comment]
@paperdigits Not being the maintainer of the project, I can't answer that definitevely... but if the goal is to have usable "versioned" forms of the database available directly from the repo, then somebody would need to set that up in a way such that those get generated from the master whenever updates are made (or otherwise when explicitly initiated, which might be safer and saner). Look in the tools directory for the related stuff.

**Comment by seebk** [comment]
Moving the DB to a separate repo for long term is the best solution for the project I think. However, this is not as as easy as it sounds as there is quite some infrastructure around the DB and where it is located. Here's a list what I guess would need to be done

- Create a new repo and copy the DB files. That's the easy part :)
- Update @bronger's calibration web service to upload requests to the new DB repo
- Move all existing calibration request items to the new DB repo issue tracker
- Update Lensfun's build scripts, tests, DB update scripts, etc and other sources to be aware of the new location. One could consider moving the DB related helper scripts to the new repo, too. 
- Develop a concept/strategy how to keep different DB versions in sync in the new repo. Currently we have the automatically created and backwards-compatible lib versions at https://wilson.bronger.org/lensfun-db/. That's also the location where the update scripts pull the data. If that is being kept, it is necessary to update this location from the update script. 
However, I would propose a more public solution where a branch version_0, version_1, .. in the DB rep is automatically updated from a CI job when new commits are added to master. But for now that's just a first idea.

In the end the problem is again that there is no active library developer. Moving the DB out of Lensfun for the short term might help but for the long run the library will loose it's connection to the DB and that might makes things even more complicated on the library side.

Having that said, if desired I can create an new repo for the DB under the Lensfun project. But i doubt that this makes sense considering the overall state of the project...

**Comment by bronger** [comment]
We have to distinguish between the versioning of the DB format and the versioning of DB releases (if any).

**Comment by bronger** [comment]
I would help doing this if the contributors agree on this step. I am mildly skeptical. I don’t see a real benefit.

**Comment by seebk** [comment]
I agree, at the current state I do not see any real benefit apart from a potential short term simplification of packaging for Darktable and other apps. The real problems of the project are not tackled by that change and they might even further complicate things. But as I am barely contributing anymore I don't want to be the one to block any progress...

**Comment by junkyardsparkle** [comment]
> The real problems of the project are not tackled by that change and they might even further complicate things.

Well, anyone who's been watching the same confusions come up over and over and over in various different places might tend to see that as one of the  "real problems". :-)

Maybe for a first small step: create lensfun/data_updates (or whatever) alongside lensfun/lensfun, with issue tracking, PRs, etc disabled (for now), with  branches for each "version flavor", containing a "version_n" subdirectory as well as a readme.md explaining which branch to use for which LensFun versions... and maybe a timestamp file? It would be good to get perspectives from packagers about the details.

This would also serve as a place to point people who want to update manually for whatever reason (for instance, there are many darktable users now for whom python scripts and tarballs are not realistically useful). Just a thought.

**Comment by bronger** [comment]
I don’t think that DB versions in branches are easier than tarballs. If lensfun-update-data is flawed, we need to improve that!

**Comment by junkyardsparkle** [comment]
Could just be different subdirectories instead of branches? The idea is to have a visible, actively updated repo *which isn't the LensFun library repo, and isn't associated with the version numbers of the library*. This would help to prevent the misperception that the way to get updated data is to "upgrade" to a higher library version number. This might be obvious to anyone familiar with the project, but not always obvious to those who are not.

Re: ways for users to get updated data, there was some discussion in a dt issue a while ago: https://github.com/darktable-org/darktable/issues/4180

**Comment by bronger** [comment]
> This would help to prevent the misperception that the way to get updated data is to "upgrade" to a higher library version number.

Ah, okay. Yes, I can understand that.

I would love to say, “I make a migration proposal” but given my current private time constraints, I cannot.  I think it is simpler than described in https://github.com/lensfun/lensfun/issues/1172#issuecomment-626316904. I particular, I think that the last point is not needed. In short, we have to crawl through tools/ and decide which links to change and which tools to move, too.

**Comment by bronger** [comment]
… but I also think that focusing all effort on making a 0.4.0 release is more important.

**Comment by eszlari** [comment]
> If lensfun-update-data is flawed, we need to improve that!

It can't be used for reproducible builds / builds that need to work without internet connection.

Maybe distribution of the database together with the application should be discouraged and applications should always use lensfun-update-data instead, but I don't know how feasible that is.

**Comment by junkyardsparkle** [comment]
> but I also think that focusing all effort on making a 0.4.0 release is more important.

It might be good to at least have the distinct repository established before 0.4.0 release, though... even if it's only a secondary location for the tarballs you're already hosting for now. It could replace the sourceforge fallback in the script.

**Comment by eszlari** [comment]
> I don’t think that DB versions in branches are easier than tarballs.

You wouldn't have to host them yourself anymore, since every branch/tag/commit can be downloaded as a zip/tar.gz from Github with a consistent URL.

**Comment by eszlari** [comment]
...not to mention the uptime uptime guarantees that Github provides.

**Comment by paperdigits** [comment]
What originally prompted this question/request is that the update tarballs have no version number and seem to change quite often.

**Comment by seebk** [comment]
If @bronger agrees I'd say in the end responsibility for a decision is in the hand of the one person who takes over responsibility and coordinates such a change and get's it going. I can't  help with the migration of the database content but if one of the regular contributors (@junkyardsparkle  ?) wants to coordinate that process and get started I can create the respective repo. If you decide to move the DB to a separate repo please make sure you have a long-term vision how to sync the respective DB format versions and keep the library tools updated.

**Comment by junkyardsparkle** [comment]
> If @bronger agrees I'd say in the end responsibility for a decision is in the hand of the one person who takes over responsibility and coordinates such a change and get's it going.

I agree that a dedicated maintainer would be needed before a full move, hopefully somebody with some insight into the needs of the packagers for various distros, as well as experience with github CI tools, etc... not me. ;) However, if the repo were created and for now used only as an additional location for the generated "ready to use" profiles, it would only require adding the syncing there to the existing process

The question then would be what additional things that repo should have in order to answer the needs of packagers...
> What originally prompted this question/request is that the update tarballs have no version number and seem to change quite often.

They have a [timestamp](https://wilson.bronger.org/lensfun-db/versions.json)... which could be integrated in a repo as tags.

**Comment by paperdigits** [comment]
Indeed they do have a timestamp, but that is inside the archive. 

For things like flatpak, NixOS, and snaps, we specific a source package to grab and a shasum to verify the archive. Since the lens profile archive is always call the same file name but the hash changes all the time, it makes it difficult to package where the shasum is required.

So with the flatpak, we were just shipping the profiles that came with the source code release, which were out of date and we don't have a great way to update them. The flatpak is sandboxes, so lensfun-update-data doesn't work.

**Comment by junkyardsparkle** [comment]
> Indeed they do have a timestamp, but that is inside the archive.

Maybe the link I posted didn't show up if you're using email to read/post here... this one:

 https://wilson.bronger.org/lensfun-db/versions.json

**Comment by paperdigits** [comment]
You're right, I didn't see that, sorry. While that might help to determine if we need to grab the package, it doesn't change what I'd written before, the hash of the tarball changes (it changes a lot actually), which makes declarative packaging difficult.

**Comment by junkyardsparkle** [comment]
Ok, so back to the question: would it be possible to solve this, for now, without changing the existing repo, but creating one alongside containing essentially the contents of those archives? If so, what would the ideal set-up look like?

**Comment by bronger** [comment]
> If @bronger agrees I'd say in the end responsibility for a decision is in the hand of the one person who takes over responsibility and coordinates such a change and get's it going.
I agree (of course).

**Comment by paperdigits** [comment]
Yes, a repo along side of the existing repo would be a good start.

I would think a folder for each version of the database would be sufficient, e.g. version_1 and version_2 folders that contain the lens profiles.

**Comment by eszlari** [comment]
The way we fix the db problem for the Flatpak of Darktable on Flathub for the time being, is by using a version of `tools/update_database/follow_db_changes.py`  that is stripped down to the conversion part.

Maybe this move of the db could be made unnecessary just by refactoring that script (put the conversion part into its own file), so people can do the conversion more easily in their own build process.

**Comment by paperdigits** [comment]
I think it'd be enough to fetch the version_1 and version_2 database files and just check them into git.
